FoodRecomendation/FoodCore.js

```diff
@@ -0,0 +1,103 @@
+const express = require('express');
+const firebaseApp =require("firebase/app");
+const router = express.Router();
+const firestore = require ("firebase/firestore");
+const {DatabaseController} = require('../common/DatabaseController');
+const {PubSubIface} = require('../common/PubSub');
+const { FoodMatcher } = require('./FoodMatcher');
+
+const firebaseApp_= firebaseApp.initializeApp({
+    apiKey: "AIzaSyAGqShV0l2hfX2eO8hYzgXnXDmpO5F0xsI",
+    authDomain: "silken-tenure-419721.firebaseapp.com",
+    projectId: "silken-tenure-419721",
+    storageBucket: "silken-tenure-419721.appspot.com",
+    messagingSenderId: "245250661585",
+    appId: "1:245250661585:web:194d2c6ccce99dd6aab9f0",
+    measurementId: "G-L0BFNM9LDG"
+  });
+
+const db = firestore.getFirestore("myrestaurant");
+
+class DishIface extends PubSubIface {
+  constructor(
+    downstream_callback, topic_name='dish', projectId = 'silken-tenure-419721'
+  ) {
+    super(topic_name, projectId);
+    this.downstream_callback = downstream_callback;
+  }
+
+  async setupTopics(topics) {
+    await super.setupTopics(topics);
+    this.downstream_sub = await this.getSubscriptionByName(this.downstream_topic, this.downstream_sub_name);
+    this.subs.push(this.downstream_sub);
+
+    this.subscribe_to_downstream(this.downstream_callback);
+  }
+
+  subscribe_to_downstream(callback) {
+    this.downstream_sub.on('message', message => callback(message));
+  }
+
+}
+
+
+//
+
+// router.get('/getMenu', async (req, res) => {
+//   try {
+//       const list = await databaseController.getAllMenu();
+//       res.send(list);
+//   } catch (error) {
+//       console.log(error)
+//       res.status(500).send(error.message);
+//   }
+// });
+
+//module.exports = router;
+
+
+class FoodCore{
+  constructor() {
+    this.dishIface = new DishIface(this.simple_callback);
+    this.databaseController = new DatabaseController();
+    this.foodMatcher = new FoodMatcher();
+  }
+
+  simple_callback  = async (message) => {
+    console.log('Received message: ', message.data.toString());
+    message.ack();
+
+   
+    const messageContent = message.data.toString();
+
+    if (messageContent === 'default message') {
+        // If the message is 'default message', do something
+        console.log('The message is the default message.');
+
+        // Now we need to return a message
+        const menu =  await this.databaseController.getAllMenuWithoutRef();
+        const jsonString = JSON.stringify(menu);
+        await this.dishIface.upstream_topic.publishMessage({data:Buffer.from(jsonString)})
+
+    } else {
+        // If the message is not 'default message', do something else
+        const parsedData = JSON.parse(messageContent);
+
+        const dish1Value = parsedData.dish1;
+        const dish2Value = parsedData.dish2;
+
+        const result = await this.foodMatcher.findMatchForDish(dish1Value, dish2Value);
```

Comment:
What happens if it fails?

Reply:
It returns to the user a 400, with the body:

SyntaxError: Expected double-quoted property name in JSON at position 25 (line 4 column 1).

Complete httpresponse would be:

<!DOCTYPE html>
<html lang="en">

<head>
    <meta charset="utf-8">
    <title>Error</title>
</head>

<body>
    <pre>SyntaxError: Expected double-quoted property name in JSON at position 25 (line 4 column 1)<br> &nbsp; &nbsp;at JSON.parse (&lt;anonymous&gt;)<br> &nbsp; &nbsp;at parse (/Users/luiscraw/Documents/TEC/SOA S1 2024/proyecto2/My_RESTaurante/BackendCore/node_modules/body-parser/lib/types/json.js:92:19)<br> &nbsp; &nbsp;at /Users/luiscraw/Documents/TEC/SOA S1 2024/proyecto2/My_RESTaurante/BackendCore/node_modules/body-parser/lib/read.js:128:18<br> &nbsp; &nbsp;at AsyncResource.runInAsyncScope (node:async_hooks:206:9)<br> &nbsp; &nbsp;at invokeCallback (/Users/luiscraw/Documents/TEC/SOA S1 2024/proyecto2/My_RESTaurante/BackendCore/node_modules/raw-body/index.js:238:16)<br> &nbsp; &nbsp;at done (/Users/luiscraw/Documents/TEC/SOA S1 2024/proyecto2/My_RESTaurante/BackendCore/node_modules/raw-body/index.js:227:7)<br> &nbsp; &nbsp;at IncomingMessage.onEnd (/Users/luiscraw/Documents/TEC/SOA S1 2024/proyecto2/My_RESTaurante/BackendCore/node_modules/raw-body/index.js:287:7)<br> &nbsp; &nbsp;at IncomingMessage.emit (node:events:519:28)<br> &nbsp; &nbsp;at endReadableNT (node:internal/streams/readable:1696:12)<br> &nbsp; &nbsp;at process.processTicksAndRejections (node:internal/process/task_queues:82:21)</pre>
</body>

</html>